src/dex/aave-v3-stata/aave-v3-stata.ts

```diff
@@ -0,0 +1,492 @@
+import { AsyncOrSync } from 'ts-essentials';
+import {
+  Token,
+  Address,
+  ExchangePrices,
+  PoolPrices,
+  AdapterExchangeParam,
+  PoolLiquidity,
+  Logger,
+  NumberAsString,
+  DexExchangeParam,
+} from '../../types';
+import { SwapSide, Network } from '../../constants';
+import * as CALLDATA_GAS_COST from '../../calldata-gas-cost';
+import { Utils, getBigIntPow, getDexKeysWithNetwork } from '../../utils';
+import { Context, IDex } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import { AaveV3StataData, Rounding, StataFunctions, TokenType } from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { AaveV3StataConfig, Adapters } from './config';
+import { Interface } from '@ethersproject/abi';
+import { fetchTokenList } from './utils';
+import {
+  getTokenFromAddress,
+  getTokenType,
+  setTokensOnNetwork,
+} from './tokens';
+import { uint256ToBigInt } from '../../lib/decoders';
+import TokenABI from '../../abi/aavev3stata/Token.json';
+import { extractReturnAmountPosition } from '../../executor/utils';
+// import { IStaticATokenLM_ABI } from '@bgd-labs/aave-address-book';
+// slimmed down version of @bgd-labs/aave-address-book
+// required as version of web3-utils used is buggy
+//import IStaticATokenFactory_ABI from '../../abi/aave-v3-stata/StaticATokenFactory.json';
+
+export const TOKEN_LIST_CACHE_KEY = 'stata-token-list';
+const TOKEN_LIST_TTL_SECONDS = 24 * 60 * 60; // 1 day
+const TOKEN_LIST_LOCAL_TTL_SECONDS = 3 * 60 * 60; // 3h
+// const RAY = BigInt(1e27);
+const RAY = BigInt(`1${'0'.repeat(27)}`);
+
+export class AaveV3Stata
+  extends SimpleExchange
+  implements IDex<AaveV3StataData>
+{
+  readonly hasConstantPriceLargeAmounts = false;
+
+  readonly isFeeOnTransferSupported = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(AaveV3StataConfig);
+
+  logger: Logger;
+
+  // static readonly stata = new Interface(IStaticATokenLM_ABI);
+  static readonly stata = new Interface(TokenABI);
+
+  private state: Record<string, { blockNumber: number; rate: bigint }> = {};
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+    protected config = AaveV3StataConfig[dexKey][network],
+    protected adapters = Adapters[network],
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+  }
+
+  // Initialize pricing is called once in the start of
+  // pricing service. It is intended to setup the integration
+  // for pricing requests. It is optional for a DEX to
+  // implement this function
+  async initializePricing(blockNumber: number) {
+    let cachedTokenList = await this.dexHelper.cache.getAndCacheLocally(
```

Comment:
no need to cache locally

Reply:
why? it's gonna be both local/distant

Reply:
because `getAndCacheLocally` with current params is called only once, therefore no need to store it locally

Reply:
done